src/hotspot/share/classfile/modules.cpp

```diff
@@ -562,13 +562,43 @@ void Modules::verify_archived_modules() {
 
 char* Modules::_archived_main_module_name = nullptr;
 char* Modules::_archived_addmods_names = nullptr;
+char* Modules::_archived_native_access_flags = nullptr;
 
 void Modules::dump_main_module_name() {
   const char* module_name = Arguments::get_property("jdk.module.main");
   if (module_name != nullptr) {
     _archived_main_module_name = ArchiveBuilder::current()->ro_strdup(module_name);
   }
-  ArchivePtrMarker::mark_pointer(&_archived_main_module_name);
+}
+
+void Modules::check_archived_flag_consistency(char** archived_flag, const char* runtime_flag, const char* property) {
+  log_info(cds)("%s %s", property,
+    *archived_flag != nullptr ? *archived_flag : "(null)");
+  bool disable = false;
+  if (runtime_flag == nullptr) {
+    if (*archived_flag != nullptr) {
+      log_info(cds)("Module for %s: %s specified during dump time but not during runtime", property, *archived_flag);
+      disable = true;
+    }
+  } else {
+    if (*archived_flag == nullptr) {
+      log_info(cds)("Module for %s: %s specified during runtime but not during dump time", property, runtime_flag);
+      disable = true;
+    } else if (strcmp(runtime_flag, *archived_flag) != 0) {
+      log_info(cds)("Mismatched modules for %s: runtime %s dump time %s", property, runtime_flag, *archived_flag);
+      disable = true;
+    }
+  }
+
+  if (disable) {
+    log_info(cds)("Disabling optimized module handling");
+    CDSConfig::stop_using_optimized_module_handling();
+  }
+  log_info(cds)("optimized module handling: %s", CDSConfig::is_using_optimized_module_handling() ? "enabled" : "disabled");
+  log_info(cds)("full module graph: %s", CDSConfig::is_using_full_module_graph() ? "enabled" : "disabled");
+
+  // Don't hold onto the pointer, in case we might decide to unmap the archive.
+  *archived_flag = nullptr;
```

Comment:
A `check_xxx` function is usually not expected to modify one of its incoming parameters. Also, `char**` variables are messy. It's better to do this in the caller:

```

check_archived_flag_consistency(_archived_main_module_name, runtime_main_module, "jdk.module.main");
check_archived_flag_consistency(_archived_native_access_flags, get_native_access_flags_as_sorted_string(), "jdk.module.enable.native.access");
...

 // Don't hold onto the pointers, in case we might decide to unmap the archive.
_archived_main_module_name = nullptr;
_archived_native_access_flags = nullptr;
```